packages/grafana-data/src/utils/dataFrameHelper.ts

```diff
@@ -0,0 +1,264 @@
+import { Field, FieldType, DataFrame, Vector, FieldDTO, DataFrameDTO } from '../types/dataFrame';
+import { Labels, QueryResultMeta } from '../types/data';
+import { guessFieldTypeForField, guessFieldTypeFromValue } from './processDataFrame';
+import { ArrayVector } from './vector';
+import isArray from 'lodash/isArray';
+
+export class DataFrameHelper implements DataFrame {
+  refId?: string;
+  meta?: QueryResultMeta;
+  name?: string;
+  fields: Field[];
+  labels?: Labels;
+  length = 0; // updated so it is the length of all fields
+
+  private fieldByName: { [key: string]: Field } = {};
+  private fieldByType: { [key: string]: Field[] } = {};
+
+  constructor(data?: DataFrame | DataFrameDTO) {
+    if (!data) {
+      data = { fields: [] }; //
+    }
+    this.refId = data.refId;
+    this.meta = data.meta;
+    this.name = data.name;
+    this.labels = data.labels;
+    this.fields = [];
+    for (let i = 0; i < data.fields.length; i++) {
+      this.addField(data.fields[i]);
+    }
+  }
+
+  addFieldFor(value: any, name?: string): Field {
+    if (!name) {
+      name = `Field ${this.fields.length + 1}`;
+    }
+    return this.addField({
+      name,
+      type: guessFieldTypeFromValue(value),
+    });
+  }
+
+  /**
+   * Reverse the direction of all fields
+   */
+  reverse() {
+    for (const f of this.fields) {
+      if (isArray(f.values)) {
+        const arr = f.values as any[];
+        arr.reverse();
+      }
+    }
+  }
+
+  private updateTypeIndex(field: Field) {
+    // Make sure it has a type
+    if (field.type === FieldType.other) {
+      const t = guessFieldTypeForField(field);
+      if (t) {
+        field.type = t;
+      }
+    }
+    if (!this.fieldByType[field.type]) {
+      this.fieldByType[field.type] = [];
+    }
+    this.fieldByType[field.type].push(field);
+  }
+
+  addField(f: Field | FieldDTO): Field {
+    const type = f.type || FieldType.other;
+    const values =
+      !f.values || isArray(f.values)
+        ? new ArrayVector(f.values as any[] | undefined) // array or empty
+        : (f.values as Vector);
+
+    // And a name
+    let name = f.name;
+    if (!name) {
+      if (type === FieldType.time) {
+        name = `Time ${this.fields.length + 1}`;
+      } else {
+        name = `Column ${this.fields.length + 1}`;
+      }
+    }
+    const field: Field = {
+      name,
+      type,
+      config: f.config || {},
+      values,
+    };
+    this.updateTypeIndex(field);
+
+    if (this.fieldByName[field.name]) {
+      console.warn('Duplicate field names in DataFrame: ', field.name);
```

Comment:
I wonder if we shouldn't throw here instead of warning. `getFieldByName` becomes not predictable unless you know the implementation.

Reply:
I don't think we can throw an exception since there is no existing contract that says field (columns) can not have the same name.

I do think it is OK to say "You will have unpredictable behavior *if* fields have the same name"

Reply:
Just added https://github.com/grafana/grafana/pull/18391/commits/3292fb717692c400c8b34765a446ad2155edfe44 this makes the behavior more clear (i hope)